examples/factory/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+//! An example demonstrating how a factory pattern can be implemented with smart
+//! contracts on Concordium.
+//!
+//! > **Important Note:** The factory pattern is not generally idiomatic on
+//! > Concordium. In many cases, a more appropriate solution is to incorporate
+//! > the factory and products into a single smart contract instance, rather
+//! > than having a separate instance for each product.
+//!
+//! # Description
+//!
+//! This example shows how a factory pattern can be implemented with smart
+//! contracts on Concordium. It consists of two smart contracts: the factory and
+//! the product.
+//!
+//! In a typical factory pattern, the factory contract provides a `produce`
+//! endpoint that constructs a new instance of the product contract. On
+//! Concordium, however, it is not possible to create a new smart contract
+//! instance within a receive method.
+//!
+//! To work around this, the new smart contract instance is first created in a
+//! separate transaction. The address of this new smart contract is then passed
+//! to the `produce` endpoint on the factory. As part of `produce`, the factory
+//! calls an `initialize` method on the product.
+//!
+//! In the typical factory pattern, the factory controls which contract is
+//! constructed and can be sure that the instance is newly constructed. In our
+//! implementation, the factory does not directly construct the product. This
+//! means that the factory **must check** that the supplied product is an
+//! instance of the expected product contract. This is achieved by checking the
+//! module reference and contract name match expected values. To ensure that the
+//! product is indeed newly constructed, the factory calls an `initialize`
+//! method on the product that will initialize its state, but fail if it has
+//! already been initialized.
+//!
+//! In this example, both the factory and product are defined in the same
+//! module, so the factory checks that the product's module reference is the
+//! same as the factory's. If the contracts were defined in separate modules,
+//! then the factory would need to know the module reference that products
+//! should use. This could be hard-coded in the contract, or set as a parameter
+//! when the factory is initialized.
+//!
+//! Since the construction and initialization of the product occur in two
+//! separate transactions, it is possible that a third party might try to hijack
+//! the process by inserting their own transaction to initialize the product
+//! (for instance, invoking a different factory instance than intended by the
+//! originator). To prevent this possibility, the product checks in its
+//! `initialize` method that the invoker of the `produce` transaction is the
```

Comment:
We try to avoid using `ctx.invoker` (`tx_origin` equivalent on Ethereum) for authorization in smart contracts.

https://consensys.github.io/smart-contract-best-practices/development-recommendations/solidity-specific/tx-origin/